src/attributes.jl

```diff
@@ -0,0 +1,280 @@
+    if !(field in propertynames(attributes))
+        # TODO, dont let anyone change propertynames!
+        push!(propertynames(attributes), field)
+        default_values = getfield(attributes, :default_values)
+        default_values[field] = value
+    end
+    return value
+end
+
+function Base.setproperty!(attributes::ObservableAttributes, field::Symbol, value::Observable)
+    setproperty!(attributes, field, value[])
+    on(value) do new_value
+        setproperty!(attributes, field, value[])
+    end
+    return value
+end
+
+function Base.setindex!(attributes::ObservableAttributes, value, field::Symbol)
+    setproperty!(attributes, field, value)
+end
+
+Base.broadcastable(x::ObservableAttributes) = Ref(x)
+
+#######
+## Dict interface
+Base.keys(x::ObservableAttributes) = propertynames(x)
+Base.values(x::ObservableAttributes) = (getproperty(x, field) for field in propertynames(x))
+function Base.pop!(x::ObservableAttributes, field::Symbol)
+    value = getproperty(x, field)
+    delete!(x, field)
+    return value
+end
+
+Base.haskey(x::ObservableAttributes, key::Symbol) = key in keys(x)
+
+Base.filter(f, x::ObservableAttributes) = ObservableAttributes(filter(f, attributes(x)))
+Base.empty!(x::ObservableAttributes) = (empty!(attributes(x)); x)
+Base.length(x::ObservableAttributes) = length(attributes(x))
+
+function Base.delete!(attributes::ObservableAttributes, field::Symbol)
+    # The priority is:
+    # User given
+    from_user = getfield(attributes, :from_user)
+    haskey(from_user, field) && delete!(from_user, field)
+    # ObservableAttributes given
+    from_theme = getfield(attributes, :from_theme)
+    haskey(from_theme, field) && delete!(from_theme, field)
+    # Construction defaults
+    default_values = getfield(attributes, :default_values)
+    haskey(default_values, field) && delete!(default_values, field)
+    delete!(propertynames(attributes), field)
+    return
+end
+
+function Base.iterate(x::ObservableAttributes, state...)
+    s = iterate(keys(x), state...)
+    s === nothing && return nothing
+    return (s[1] => x[s[1]], s[2])
+end
+
+function Base.copy(attributes::ObservableAttributes)
+    return ObservableAttributes(attributes)
+end
+
+Base.merge(target::ObservableAttributes, args::ObservableAttributes...) = merge!(copy(target), args...)
+
+function merge_attributes!(input::ObservableAttributes, theme::ObservableAttributes)
+    for (key, value) in theme
+        if !haskey(input, key)
+            input[key] = value
+        else
+            current_value = input[key]
+            if value isa ObservableAttributes && current_value isa ObservableAttributes
+                # if nested attribute, we merge recursively
+                merge_attributes!(current_value, value)
+            elseif value isa ObservableAttributes || current_value isa ObservableAttributes
+                error("""
+                Type missmatch while merging plot attributes with theme for key: $(key).
```

Comment:
``suggestion
                Type mismatch while merging plot attributes with theme for key: $(key).